### ISSUE TITLE
[2018]  Consistent Individualized Feature Attribution for Tree Ensembles

### ISSUE THREAD

**Issue by yoheikikuta** [yoheikikuta/paper-reading#17]
## 論文リンク
https://arxiv.org/abs/1802.03888

## 公開日（yyyy/mm/dd）
2018/02/12

## 概要
機械学習の解釈可能性に関する SHAP 値を XGBoost や LightGBM に対して効率的に計算したり実験で確認したという論文。
詳しくはあとで書く。

**Comment by yoheikikuta** [comment]
GitHub repository https://github.com/slundberg/shap にリンクが貼られていて、ちょっと SHAP あたりを勉強してみようかなと思って読み始めてみる。

元々の話は ICML workshop の https://arxiv.org/abs/1706.06060 の論文で、その続き物という感じのようだ。

ちなみに SHAP の original paper は https://arxiv.org/abs/1705.07874 の論文。こっちを読む方が自分が知りたいことが知れるのではないかという気もするが、まあ必要がありそうだったら読んでみることにしよう。

**Comment by yoheikikuta** [comment]
tree ensemble でよく使う変数の寄与度としては importance がある。
これは全データに対して考えて学習したモデルの変数の importance を考える場合と（特徴量選定などでよく使うのはこっち）、個別のデータに対する予測を root から leaf までの path で各ノードでの寄与を考える場合がある。

話のステップとしてはざっくり以下のような感じかな。
- よく使われてきたような importance は "inconsistent" であることを説明
- consistent な方法として SHAP 値による定式化
- efficient な計算アルゴリズムを提示
- 実験で確認

**Comment by yoheikikuta** [comment]
inconsistency の説明は以下の図でなされる。
一言で言えば、モデルを変更した時（イメージとしては決定木のノードで使う feature を変更）にある feature の予測への寄与がより大きくなるような場合に importance が下がってしまう、ということである。

これが起こる状況を理解するために以下の図を眺める。これは simple な regression tree で風である percentage を予測するものに見える。
Model A では Fever が root node にあり、child node に Cough があり、棒グラフの左側の結果が得られている。とのことだが具体的にどう計算しているのかってのがちょっとよくわからない。イメージ図ってことなのだろうか。二段目の split で {0,0} になっているのになぜか分かれていて「？」となったりする（これは小さい数を丸めてるからと注釈があるが）。
まあ言いたいことはこのような Model A があるときに次に Model B が得られた場合に inconsistent だということなのでそこだけは理解しておく。Model B は Cough が root node になるので、Cough がより重要になっていると考える。しかし importance の各指標を計算すると、Fever の方が高くなって Model A と比べると逆転するので、その意味で inconsistent ということである。
ちょっと混乱するが、あくまで Model A をベースとして見た時に、Model B でより重要となった feature の importance が減少してしまうという話だ。最近提案された SHAP であればそのような inconsistency が起きてないことがポイントで、なぜそうなるかは先を読んでいくと分かるはず。

<p align="center">
  <img src="https://i.imgur.com/e6aD8Id.png">
</p>

**Comment by yoheikikuta** [comment]
次は SHAP の review part が続く。

重要な additive feature attribution methods とは、以下のように explanation model を作れるというものだ。

<p align="center">
  <img src="https://i.imgur.com/dMl7520.png">
</p>

単純な linear model で説明したい関数 f(x) を近似するのが目的だ。φ が各 feature の attribution を示すものになっている。z が {0,1} で、当該の feature のあるなしを flag として持つものになっている。なので、登場する feature (z=1) の attribution の和というシンプルな形で f(x) を近似する。

こいつが game theory で single unique solution が存在することが示されるという重要な結果は先行研究で示されているので、そこは今度興味が出たら読んでみよう。先行研究で示されている重要な特徴は以下三つらしい。
- local accuracy: feature attribution の和が説明したい関数の output と等しい f(x) = Σ φ_i
- missingness: z=0 の feature は寄与しない（これは定式化から当然な気がするがより深い意味があるのか？） 
- consistency: 上で説明した話。モデルを変更してより大きなインパクトを与える feature の attribution はその変更で減少しない。

理論的に single unique solution の存在が示せるというのは面白いね。というかそれが shapley value の出典なんですね。どこかでまた出会ったら勉強しよう。

**Comment by yoheikikuta** [comment]
SHAP value の計算方法は以下。

<p align="center">
  <img src="https://i.imgur.com/o7fMjPq.png">
</p>

雑に言ってしまえば、ある feature の存在を index として S で表し、その S に他の feature を {i} として追加した時のモデルの output 期待値の差分を全 feature の組み合わせで計算するというものになっている。

ある feature を入れた時に出力値がどれくらい変わるか、というので attribution としていることが分かる。全部の組み合わせを入れてるので、その意味で相互作用も考慮できているのだろう。熱がある feature による出力値も、咳が出るという feature がある上での熱がある feature による出力値も考慮しているので。そのような複雑な組み合わせの効果はギュッと φ に押し込められ、その φ を用いて liear model で元関数を近似するというわけだな。

**Comment by yoheikikuta** [comment]
この計算は指数時間を要求する。
それを tree 構造をうまく活かして木の深さの二乗で抑えた、しかし exact な、アルゴリズムを提唱したというのがこの論文の main contribution である。

これは直感的には木構造の subset の情報を track しておいてそれを combine したりすることで効率的に計算するという感じだろうか。
アルゴリズムをちゃんと理解するには結構腰を据えて読む必要がありそうなのだが、自分はここのアルゴリズム自体をちゃんと理解したいというモチベーションで読み始めたわけではないので、将来知りたくなったらまた読むとして論文の Algorithm 2 はそういうものがあるらしいということで読み進めることにする。
main contribution がここなのでは！？という心の声は一旦置いておいて、SHAP まわりを軽く勉強して見たいと思っていた最初のモチベーションの下で先に進めることにする。

**Comment by yoheikikuta** [comment]
もう一つ重要な量として、SHAP interaction value を導入している。
気持ちとしては、上で計算した SHAP value はあらゆる相互作用を含んだ結果になっているので、そこから二つの組み合わせに注目してその interaction の SHAP value を計算することで、単体の効果と interaction の効果を観察してより深い洞察が得たいという感じだ。

<p align="center">
  <img src="https://i.imgur.com/EYv0Y0n.png">
</p>

基本的には単体の SHAP value の計算からの自然な拡張なので特に大変なところはないだろう。
重要な性質として $` i \leftrightarrow j `$の symmetry がある。
元の SHAP value から interaction SHAP を差っ引くことで、main contribution の効果を読み取れるだろうという算段である。

**Comment by yoheikikuta** [comment]
最後に最高血圧のデータに対して、main effect と interaction を分けて SHAP 値を計算する実験。

<p align="center">
  <img src="https://i.imgur.com/VZ2cYsO.png">
</p>

上の方が main effect の結果である。最高血圧は低ければ SHAP 値はゼロで、120くらいならば死亡のリスクを減らす方向に働き、140を超えたあたりから高くなるほど死亡のリスクが高くなることを示している。他の feature との interaction の効果を全て覗いてるので、血圧値一定で縦方向を見るとほとんど変化がなく、確かに血圧値の効果だけが記述できていることが見て取れる。

下の方は interaction のうち最高血圧と年齢のものを取り出した結果である。最高血圧が140を超えたあたりから、年齢が若い方が血圧高いと死亡のリスクが高く、老人であれば死亡のリスクが低い、という結果になっている。老人は比較的平均血圧が高いのでこのような結果になっているということだろう。

**Comment by yoheikikuta** [comment]
ということで SHAP を ensemble tree に適用するという論文を読んでみた。
肝心のアルゴリズムの部分はちゃんと理解していないが、SHAP のことが表面的には理解できるようになってきたので最初の目標は達成できた。結構流行りのトピックでもあるし、どこかで遭遇したらまた少し理解を深めていくことにしよう。
